### PR TITLE
[JioSaavn] Add extractor

### PR DESCRIPTION
**Pull request by awalgarg**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/ytdl-org/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Extractor for Saavn/JioSaavn/JioMusic (see https://en.wikipedia.org/wiki/JioSaavn).

While all the urls from the original saavn.com website now redirect to the new website, there are megatons of them in google's cache, bookmarks and such - and thus the regex in this PR accounts for those old urls as well.

### CODE REVIEW

File: youtube_dl/extractor/jiosaavn.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+            'album': 'Saathiya',
+            'release_year': 2002,
+        },
+    }]
+
+    def _real_extract(self, url):
+        audio_id = self._match_id(url)
+
+        _fp = ''.join([random.choice(string.hexdigits) for _ in range(32)])
+        self._download_webpage('https://www.jiosaavn.com/stats.php?ev=site:browser:fp&fp=%s&_t=%d&ct=%s' % (_fp, int(time.time()), '00000000'), audio_id)
```
Comment (dstftw) [inline youtube_dl/extractor/jiosaavn.py]: Move query to `query`.

Comment (dstftw) [inline youtube_dl/extractor/jiosaavn.py]: There should be a comment pointing out the need of downloading this.

File: youtube_dl/extractor/jiosaavn.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+        webpage = self._download_webpage(url, audio_id)
+
+        song_data = self._parse_json(self._search_regex(r'(?s)<div class="hide song-json">(?P<json>.+?)\</div>',
+                                                        webpage, 'song-json', group='json'),
```
Comment (dstftw) [inline youtube_dl/extractor/jiosaavn.py]: There is no need in named group when it's the only one.

File: youtube_dl/extractor/jiosaavn.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        'md5': 'd67db3da18f4e72c5558b213d6698704',
+        'info_dict': {
+            'id': 'OQsEfQFVUXk',
+            'ext': 'mp3',
+            'title': 'Leja Re',
+            'album': 'Leja Re',
+            'release_year': 2018,
+        },
+    }, {
+        'url': 'https://www.saavn.com/s/song/hindi/Saathiya/O-Humdum-Suniyo-Re/KAMiazoCblU',
```
Comment (dstftw) [inline youtube_dl/extractor/jiosaavn.py]: Make `only_matching`.

**Comment by awalgarg** [comment]
Addressed review comments

**Comment by awalgarg** [comment]
@dstftw *gentle nudge*

File: youtube_dl/extractor/jiosaavn.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
+# coding: utf-8
+from __future__ import unicode_literals
+
+import random
+import string
+import itertools
+import time
```
Comment (dstftw) [inline youtube_dl/extractor/jiosaavn.py]: Alphabetic.

File: youtube_dl/extractor/jiosaavn.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+from .common import InfoExtractor
+from ..utils import (
+    int_or_none,
+    urlencode_postdata,
+    unescapeHTML,
```
Comment (dstftw) [inline youtube_dl/extractor/jiosaavn.py]: Alphabetic.

File: youtube_dl/extractor/jiosaavn.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+        webpage = self._download_webpage(url, audio_id)
+
+        song_data = self._parse_json(self._search_regex(r'(?s)<div class="hide song-json">(?P<json>.+?)\</div>',
+                                                        webpage, 'song-json'),
+                                     audio_id, transform_source=unescapeHTML)
+
+        title = song_data['title']
+
+        def clean_api_json(resp):
+            return '\n'.join(itertools.dropwhile(lambda line: not line.startswith('{'), resp.split('\n')))
```
Comment (dstftw) [inline youtube_dl/extractor/jiosaavn.py]: Just `_search_regex`.

File: youtube_dl/extractor/jiosaavn.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+                               query={
+                                   'ev': 'site:browser:fp',
+                                   'fp': ''.join([random.choice(string.hexdigits) for _ in range(32)]),
+                                   '_t': int(time.time()),
+                                   'ct': '00000000',
+                               })
+
+        webpage = self._download_webpage(url, audio_id)
+
+        song_data = self._parse_json(self._search_regex(r'(?s)<div class="hide song-json">(?P<json>.+?)\</div>',
```
Comment (dstftw) [inline youtube_dl/extractor/jiosaavn.py]: Carry arguments to the next lines not to waste space on the left. Keep 80 char len per line.

File: youtube_dl/extractor/jiosaavn.py [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+                                                        webpage, 'song-json'),
+                                     audio_id, transform_source=unescapeHTML)
+
+        title = song_data['title']
+
+        def clean_api_json(resp):
+            return '\n'.join(itertools.dropwhile(lambda line: not line.startswith('{'), resp.split('\n')))
+
+        media_url = self._download_json('https://www.jiosaavn.com/api.php', audio_id,
```
Comment (dstftw) [inline youtube_dl/extractor/jiosaavn.py]: Same.

**Comment by awalgarg** [comment]
@dstftw All fixed

**Comment by awalgarg** [comment]
Bump

@dstftw

**Pull request closed without merging**